samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -285,34 +297,41 @@ void handleExpiredRequest(String processorId, String preferredHost,
    * @param containerAllocator to request physical resources
    */
   public void registerContainerPlacementAction(ContainerPlacementRequestMessage requestMessage, ContainerAllocator containerAllocator) {
-    LOG.info("Received a ContainerPlacement action request: {}", requestMessage);
     String processorId = requestMessage.getProcessorId();
     String destinationHost = requestMessage.getDestinationHost();
-    Pair<ContainerPlacementMessage.StatusCode, String> actionStatus =
-        checkValidControlAction(processorId, destinationHost, requestMessage.getUuid());
-
+    Optional<Pair<ContainerPlacementMessage.StatusCode, String>> actionStatus = validatePlacementAction(requestMessage);
+    if (!actionStatus.isPresent()) {
+      // Action is supposed to be queued
+      LOG.info("ContainerPlacement request is en-queued metadata: {}", requestMessage);
+      return;
+    }
+    LOG.info("ContainerPlacement action is de-queued metadata: {}", requestMessage);
+    // Remove the request message from metastore since this message is already acted upon
+    containerPlacementMetadataStore.deleteContainerPlacementRequestMessage(requestMessage.getUuid());
+    // Action is de-queued upon so we record it in the cache
+    dequeuedActionsCache.put(requestMessage.getUuid());
     // Request is bad just update the response on message & return
-    if (actionStatus.getKey() == ContainerPlacementMessage.StatusCode.BAD_REQUEST) {
+    if (actionStatus.get().getKey() == ContainerPlacementMessage.StatusCode.BAD_REQUEST) {
+      writeContainerPlacementResponseMessage(requestMessage, actionStatus.get().getKey(), actionStatus.get().getValue());
       return;
     }
 
     SamzaResource currentResource = samzaApplicationState.runningProcessors.get(processorId);
-    LOG.info(
-        "Processor ID: {} matched an active container with deployment ID: {} is running on host: {} for ContainerPlacement action: {}",
+    LOG.info("Processor ID: {} matched an active container with deployment ID: {} is running on host: {} for ContainerPlacement action: {}",
         processorId, currentResource.getContainerId(), currentResource.getHost(), requestMessage);
 
     if (!hostAffinityEnabled) {
-      LOG.info("Changing the requested host for placement action to {} because host affinity is disabled",
-          ResourceRequestState.ANY_HOST);
+      LOG.info("Changing the requested host for placement action to {} because host affinity is disabled", ResourceRequestState.ANY_HOST);
       destinationHost = ANY_HOST;
     }
```

Comment:
Is there a specific need for this?

Reply:
"Shoudnt ANY_HOST be the default, while if I want to be able to send a stateless job's container to a specific host, it should rather be allowed."

That needs some additional work in AM because ResourceRequestStatue and other parts of AM have significant code that just checks for this config and identifies a preferred resource as ANY_HOST, all of that needs to be refactored